Students/Truta Andrei/Lab05/src/ro/utcn/Customer.java

```diff
@@ -0,0 +1,121 @@
+        this.id = ++nextId;
+        this.name = name;
+        this.address = address;
+        this.email = email;
+        this.phone = phone;
+    }
+
+    public void openAccount(BankAccount.AccountType accountType) {
+        BankAccount account = new BankAccount(accountType);
+        accounts.add(account);
+    }
+
+    private BankAccount getAccount(Integer accountNumber) {
+        for (BankAccount account : accounts) {
+            if (account.getAccountNumber().equals(accountNumber)) {
+                return account;
+            }
+        }
+        return null;
+    }
+
+    public void depositCurrency(Integer accountNumber, Double amount) {
+        BankAccount account = getAccount(accountNumber);
+
+        if (account == null) {
+            throw new IllegalArgumentException("Invalid account number");
+        }
+
+        account.depositCurrency(amount);
+    }
+
+    public void withdrawCurrency(Integer accountNumber, Double amount) {
+        BankAccount account = getAccount(accountNumber);
+
+        if (account == null) {
+            throw new IllegalArgumentException("Invalid account number");
+        }
+
+        account.withdrawCurrency(amount);
+    }
+
+    //Moves currency between own accounts
+    public void transferInternalCurrency(Integer sourceAccountNumber, Integer destinationAccountNumber, Double amount) {
+        BankAccount sourceAccount = getAccount(sourceAccountNumber);
+        BankAccount destinationAccount = getAccount(destinationAccountNumber);
+
+        if (sourceAccount == null || destinationAccount == null) {
+            throw new IllegalArgumentException("Invalid account number");
+        }
+
+        Transaction transaction = new Transaction(sourceAccount, destinationAccount, Transaction.TransactionType.TRANSFER_OUT, amount, null);
+        sourceAccount.addTransaction(transaction);
+
+        transaction = new Transaction(sourceAccount, destinationAccount, Transaction.TransactionType.TRANSFER_IN, amount, null);
+        destinationAccount.addTransaction(transaction);
+    }
+
+    //Transfer between accounts of different customers
+    public void transferExternalCurrency(Integer sourceAccountNumber, Customer destinationCustomer, Integer destinationAccountNumber, Double amount) {
+        BankAccount sourceAccount = getAccount(sourceAccountNumber);
+        BankAccount destinationAccount = destinationCustomer.getAccount(destinationAccountNumber);
+
+        if (sourceAccount == null || destinationAccount == null) {
+            throw new IllegalArgumentException("Invalid account number");
+        }
+
+        //TODO: If I handle the invalid input here, addTransaction() can assume that the input is valid
+        //TODO: Should I add checks in addTransaction() as well? I see no reason because it's only called from here
+        //TODO: But having a method that doesn't validate the input seems weird
+        if (amount > sourceAccount.getBalance()) {
```

Comment:
single responsibility principle. Cineva trebuie sa fie responsabil de validarea amount-ului ce urmeaza a fi tranzactionat. Cel mai bine e sa fie acolo unde se cunoaste exact informatia despre current amount si amount-ul ce urmeaza a fi scos de acolo. Daca faci check-ul asta in mai multe locuri nu-i neaparat o problema.. doar daca trebuie modificat check-ul e nasol. Important e sa indeplinesti prima chestie ce-am zis-o aici:D